Assert: Introduce timeout to set per-test timeout durations

trentmwillis opened the pull request:
As discussed in https://github.com/qunitjs/qunit/issues/1132.

This makes it easy to set per-test timeouts, instead of the globally configured `QUnit.config.testTimeout`.

Krinkle commented on docs/assert/timeout.md:
```diff
@@ -0,0 +1,36 @@
+
+### Description
+
+`assert.timeout()`sets the length of time, in milliseconds, to wait for async operations in the current test. This is equivalent to setting `config.testTimeout` on a per-test basis. The timeout length only applies when performing async operations.
```
Missing space between "timeout()" and "sets".

Krinkle commented on src/assert.js:
```diff
@@ -14,6 +14,14 @@ class Assert {
 	// Assert helpers
 
+	timeout( duration ) {
+		if ( objectType( duration ) !== "number" ) {
```
I don't think we need objectType here. `typeof` should suffices, right?

trentmwillis replied:
For 99% of cases it's probably fine, but if someone were to pass a `new Number(1000)`, we'd miss it, even though it is a `number`.

Krinkle replied:
I don't think we should support that. I'd rather have it throw an exception. These are not wild values, they are crafted for the specific purpose of setting a timeout in QUnit.

Krinkle commented on src/test.js:
```diff
@@ -36,6 +36,7 @@ export default function Test( settings ) {
 	this.module = config.currentModule;
 	this.stack = sourceFromStacktrace( 3 );
 	this.steps = [];
+	this.timeout = 0;
```
`config.testTimeout` is undefined by default. Having this one be `0` feels inconsistent. On the other hand, I don't like undefined as default for `config.testTimeout`. `false` is probably a better fit here. `0` could also work although my first-impression expectation based on other software would be that 0 either means infinite waiting or that no waiting is allowed (e.g. any async will be considered an error). Whereas in our case it means no waiting. Although I suppose infinite waiting and no waiting are very similar in the case of qunit - except that issue #1132 exposes a difference, albeit through a bug we should fix separately.

If we go with `0` we should document that for this and for `config.testTimeout` and which ever value we pick, we should file an issue to also change `config.testTimeout`'s default in the next major release.

gibson042 replied:
I agree that `0` is a poor default (it should describe the lowest meaningful timeout value), but I don't mind `undefined` for "no test-specific timeout".

trentmwillis replied:
I didn't put much thought into this when I was implementing it originally. `undefined` seems reasonable and is consistent with the current `testTimeout`, so I will go that route.

platinumazure commented:
@trentmwillis I'm going to wait to see what shakes out of Timo's review before reviewing this myself. Feel free to ping me when you've committed any changes.

gibson042 commented on docs/assert/timeout.md:
```diff
@@ -0,0 +1,36 @@
+  var done = assert.async();
+  var input = $( "#test-input" ).focus();
+  setTimeout(function() {
+    assert.equal( document.activeElement, input[0], "Input was focused" );
```
Could we have a non-DOM example?

Krinkle replied:
Agree with @gibson042 here. Note: Since the above comment was added, a non-DOM example was also added below, which is great.

I don't particularly mind having an example involving DOM in someway, but as unimportant as it might seem, the specific example we have above seems like an anti-pattern I'd rather not promote in the documentation.

I'd be okay with having just the below example, although in that case I'd rather have the main example showcase `assert.async()` which is easier to understand when learning QUnit. Having two examples, where one shows compat with Promise still would be nice too, though.

gibson042 commented on src/test.js:
```diff
@@ -656,20 +657,23 @@ export function only( testName, callback ) {
 	// Set a recovery timeout, if so configured.
-	if ( config.testTimeout && defined.setTimeout ) {
+	const timeoutDuration = test.timeout || config.testTimeout;
+	if ( timeoutDuration && defined.setTimeout ) {
```
Per the above, this should be more explicit. There's also a question of whether the global value is a fallback or a _maximum_. I'm personally inclined towards the latter, e.g.
```js
const specificTimeout = typeof test.timeout === "number" ? test.timeout : Infinity;
const globalTimeout = typeof config.testTimeout === "number" ? config.testTimeout : Infinity;
const resolvedTimeout = Math.min( explicitTimeout, globalTimeout );
if ( resolvedTimeout < Infinity && defined.setTimeout ) {
	…
}
```

trentmwillis replied:
I prefer the fallback approach. It is conceivable to want the majority of your tests to use a short timeout and then override that for a specific, slow test. If we go with the global maximum approach, we force the default for all tests to be the longer value.

trentmwillis commented:
@Krinkle @gibson042 I've updated and believe I have addressed all comments. Another review would be appreciated 😄

trentmwillis commented:
@qunitjs/qunit-team can I get someone to review this? I'd like to land this before releasing `2.4`. Thanks!

Krinkle commented on test/main/assert/timeout.js:
```diff
@@ -0,0 +1,61 @@
+QUnit.test( "does not push a failure if test is synchronous", function( assert ) {
+	assert.timeout( 1 );
+
+	var wait = Date.now() + 50;
```
Let's use `10` here as with the other one. Will make the test run a bit faster, too.

trentmwillis replied:
Rebased + updated.

trentmwillis commented on docs/assert/timeout.md:
```diff
@@ -0,0 +1,46 @@
+
+`assert.timeout()` sets the length of time, in milliseconds, to wait for async operations in the current test. This is equivalent to setting `config.testTimeout` on a per-test basis. The timeout length only applies when performing async operations.
+
+If `0` is passed, then the test will be assumed to be completely synchronous. If a non-numeric value is passed as an argument, the function will throw an error.
```
To summarize (to ensure I'm understanding everything):
* `setTimeout(, 0)` doesn't guarantee synchronicity (agreed)
* However, we _do want_ to guarantee synchronicity when using a timeout of `0`
* Therefore, we should special case `assert.timeout(0)`
* And, add tests handling edge cases that likely fail with the current implementation
Does that seem correct?

Krinkle replied:
Yep, that's right.

trentmwillis commented:
I have updated and addressed @Krinkle's concerns.

@platinumazure I've disabled the `max-len` ESLint rule for the `test` directory since it is not uncommon for test descriptions to be longer than it and I feel it adds negative value to split them on multiple lines.

Krinkle commented on test/main/assert/timeout.js:
```diff
@@ -0,0 +1,101 @@
+			stubPushFailure( assert );
+
+			var done = assert.async();
+			setTimeout( () => done(), 0 );
```
(Causes CI failure) Too new syntax for the web browser we use for our Travis tests, and presumably for some of our other browsers as well.

trentmwillis replied:
Bah, a side-effect of too much Babel usage...

trentmwillis commented:
CI passing again.

Krinkle commented on src/test.js:
```diff
@@ -164,6 +165,15 @@ Test.prototype = {
+
+			// If the test has a "lock" on it, but the timeout is 0, then we push a
+			// failure as the test should be synchronous.
+			if ( test.timeout === 0 && test.semaphore !== 0 ) {
```
I keep staring at `test.semaphore !== 0` here, and below at `expect( 1 )` in the failing test cases. I can't quite put my finger on it... Are they needed? What are they for in this case?

platinumazure commented:
@trentmwillis re: max-len Sounds good to me.
